### DISCUSSION TITLE
Warn user when using .tsx instead of .ts for API routes

### DISCUSSION THREAD

**Discussion by tomelliot** [vercel/next.js#57252] [Ideas]
### Goals

1. A build-time warning when a route isn't using the correct extension.
2. Help the developer realise when they've made a mistake in naming their API routes.
3. Prevent wasted time trying to debug issues (e.g. Sentry not auto-instrumenting API routes).

### Non-Goals

1. Change in run-time functionality

### Background

I was having an issue with Sentry not autoinstrumenting API endpoints. Because this was a recent feature release from Sentry (to enable it with the App Router), I wasn't sure where the issue was originating (Sentry's new work, my migration to use the updated Sentry, or something else).

Next.js made no complaint about me using .tsx file extension for API routes. Sentry wouldn't autoinstrument them though.

### Proposal

A warning at build time if API routes are using the wrong file extension.

**Comment by lforst** [comment]
Sentry Next.js SDK maintainer here. We took the current version of the Next.js docs too literal and didn't auto instrument `route.tsx` files - we will update the SDK so it does.

![Screenshot 2023-10-23 at 12 35 37](https://github.com/vercel/next.js/assets/8118419/5814723a-71b7-4e09-8332-2ae817b0563e)

https://nextjs.org/docs/app/building-your-application/routing/route-handlers#convention

Additionally, we should probably also change the docs stating that `.tsx` is allowed? Not a must of course if the convention should be `.ts`.

**Comment by Dragate** [comment]
Sometimes .jsx|tsx makes sense for an API route. It's not uncommon that an email has to be sent containing markup constructed from react components when the route is invoked.
